impl/src/main/java/org/ehcache/internal/persistence/DefaultLocalPersistenceService.java

```diff
@@ -138,34 +138,32 @@ static void createLocationIfRequiredAndVerify(final File rootDirectory) {
   public FileBasedPersistenceContext createPersistenceContext(Object identifier, Store.PersistentStoreConfiguration<?, ?, ?> storeConfiguration) throws CachePersistenceException {
     String stringIdentifier = validateIdentifier(identifier);
     DefaultFileBasedPersistenceContext context = new DefaultFileBasedPersistenceContext(getDirectoryFor(stringIdentifier));
-    knownPersistenceContexts.put(stringIdentifier, context);
+    if (knownPersistenceContexts.putIfAbsent(stringIdentifier, context) == null) {
+      if (!storeConfiguration.isPersistent()) {
+        destroy(stringIdentifier, context, false);
+      }
 
-    if (!storeConfiguration.isPersistent()) {
-      destroy(stringIdentifier, context, false);
-    }
+      try {
+        create(context);
+      } catch (IOException e) {
+        knownPersistenceContexts.remove(stringIdentifier, context);
+        throw new CachePersistenceException("Unable to create persistence context for " + identifier, e);
+      }
 
-    try {
-      create(context);
-    } catch (IOException e) {
-      knownPersistenceContexts.remove(stringIdentifier);
-      throw new CachePersistenceException("Unable to create persistence context for " + identifier, e);
+      return context;
+    } else {
+      throw new CachePersistenceException("Identifier '" + identifier + "' already created by this persistence service");
     }
-
-    return context;
   }
 
   @Override
-  public void destroyPersistenceContext(Object identifier) {
+  public void destroyPersistenceContext(Object identifier) throws CachePersistenceException {
     String stringIdentifier = validateIdentifier(identifier);
-    FileBasedPersistenceContext persistenceContext = knownPersistenceContexts.get(stringIdentifier);
-    if (persistenceContext != null) {
-      try {
-        destroy(stringIdentifier, persistenceContext, true);
-      } finally {
-        knownPersistenceContexts.remove(stringIdentifier);
-      }
+    FileBasedPersistenceContext persistenceContext = knownPersistenceContexts.remove(stringIdentifier);
+    if (persistenceContext == null) {
+      throw new CachePersistenceException("Identifier " + identifier + " not known to this persistence service");
```

Comment:
We cannot do that, because we want the following scenario to be possible:

1. Create a persistent cache "foo" and fill it
2. Shut down VM
3. Restart VM, initialise cache manager without initialising caches
4. Destroy cache "foo"'s persistence context before initialising it.

And I believe at that point, the context is not marked as known.